terraform/transform_deposed.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	// Go through all the resources in our state to look for tainted resources
+	for k, rs := range state.Resources {
+		if rs.Deposed == nil {
+			continue
+		}
+
+		g.Add(&graphNodeDeposedResource{
+			ResourceName: k,
+			ResourceType: rs.Type,
+		})
+	}
+
+	return nil
+}
+
+// graphNodeDeposedResource is the graph vertex representing a deposed resource.
+type graphNodeDeposedResource struct {
+	ResourceName string
+	ResourceType string
+}
+
+func (n *graphNodeDeposedResource) Name() string {
+	return fmt.Sprintf("%s (deposed)", n.ResourceName)
+}
+
+func (n *graphNodeDeposedResource) ProvidedBy() []string {
+	return []string{resourceProvider(n.ResourceName)}
+}
+
+// GraphNodeEvalable impl.
+func (n *graphNodeDeposedResource) EvalTree() EvalNode {
+	var provider ResourceProvider
+	var state *InstanceState
+
+	seq := &EvalSequence{Nodes: make([]EvalNode, 0, 5)}
+
+	// Build instance info
+	info := &InstanceInfo{Id: n.ResourceName, Type: n.ResourceType}
+	seq.Nodes = append(seq.Nodes, &EvalInstanceInfo{Info: info})
+
+	// Refresh the resource
+	seq.Nodes = append(seq.Nodes, &EvalOpFilter{
+		Ops: []walkOperation{walkRefresh},
+		Node: &EvalSequence{
+			Nodes: []EvalNode{
+				&EvalGetProvider{
+					Name:   n.ProvidedBy()[0],
+					Output: &provider,
+				},
+				&EvalReadState{
+					Name:    n.ResourceName,
+					Deposed: true,
+					Output:  &state,
+				},
+				&EvalRefresh{
+					Info:     info,
+					Provider: &provider,
+					State:    &state,
+					Output:   &state,
+				},
+				&EvalWriteState{
+					Name:         n.ResourceName,
+					ResourceType: n.ResourceType,
+					State:        &state,
+					Deposed:      true,
+				},
+			},
+		},
+	})
+
+	// Apply
+	var diff *InstanceDiff
+	var err error
+	var emptyState *InstanceState
+	tainted := true
+	seq.Nodes = append(seq.Nodes, &EvalOpFilter{
+		Ops: []walkOperation{walkApply},
+		Node: &EvalSequence{
+			Nodes: []EvalNode{
+				&EvalGetProvider{
+					Name:   n.ProvidedBy()[0],
+					Output: &provider,
+				},
+				&EvalReadState{
+					Name:    n.ResourceName,
+					Deposed: true,
+					Output:  &state,
+				},
+				&EvalDiffDestroy{
+					Info:   info,
+					State:  &state,
+					Output: &diff,
+				},
+				&EvalApply{
+					Info:     info,
+					State:    &state,
+					Diff:     &diff,
+					Provider: &provider,
+					Output:   &state,
+					Error:    &err,
+				},
+				// Always write the resource back to the state tainted... if it
+				// successfully destroyed it will be pruned. If it did not, it will
+				// remain tainted.
+				&EvalWriteState{
+					Name:         n.ResourceName,
+					ResourceType: n.ResourceType,
+					State:        &state,
+					Tainted:      &tainted,
+					TaintedIndex: -1,
+				},
+				// Then clear the deposed state.
+				&EvalWriteState{
+					Name:         n.ResourceName,
+					ResourceType: n.ResourceType,
+					State:        &emptyState,
+					Deposed:      true,
+				},
+				&EvalReturnError{
+					Error: &err,
+				},
```

Comment:
It sounds like this will work properly now... there might still be a race where we're adding a tainted and a dynamic tainted node is being made, but as you said if its nil nothing will happen.

One case to think about: is it possible to race, fail a destroy, mark it as tainted, then have the tainted node pick it up and try a _second_ destroy. I think we should avoid this.

The easiest thing might just be to just write it back to the deposed.